### ISSUE TITLE
nixos-21.11: perlPackages.ImageMagick fails to build (Shutter’s dependency)

### ISSUE THREAD

**Issue by unclechu** [NixOS/nixpkgs#156826]
### Describe the bug

`perlPackages.ImageMagick` fails to build. Couldn’t update my NixOS because Shutter depends on this library.

### Steps To Reproduce

Have this `nixos` channel pin:

https://releases.nixos.org/nixos/21.11/nixos-21.11.335481.604c44137d9 (release date is 2022-01-24T11:27:17Z)

``` console
$ nix-shell -p perlPackages.ImageMagick
these derivations will be built:
  /nix/store/f8ygrk39l0lgy3s8vs5pwlxgs1af09qa-perl5.34.0-Image-Magick-7.0.11-1.drv
these paths will be fetched (0.06 MiB download, 0.30 MiB unpacked):
  /nix/store/q6ayjwizl0xcbvknkcksy5g4n1ipa6b1-bash-interactive-5.1-p8-dev
copying path '/nix/store/q6ayjwizl0xcbvknkcksy5g4n1ipa6b1-bash-interactive-5.1-p8-dev' from 'https://cache.nixos.org'...
building '/nix/store/f8ygrk39l0lgy3s8vs5pwlxgs1af09qa-perl5.34.0-Image-Magick-7.0.11-1.drv'...
unpacking sources
unpacking source archive /nix/store/z0gw4b4nrvkwxwc303chcya8k7i0xc0b-Image-Magick-7.0.11-1.tar.gz
source root is Image-Magick-7.0.11
setting SOURCE_DATE_EPOCH to timestamp 1619956237 of file Image-Magick-7.0.11/demo/Makefile
patching sources
configuring
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
patching ./demo/annotate_words.pl...
patching ./demo/annotate.pl...
patching ./demo/button.pl...
patching ./demo/shadow-text.pl...
patching ./demo/composite.pl...
patching ./demo/settings.pl...
patching ./demo/compose-specials.pl...
patching ./demo/single-pixels.pl...
patching ./demo/piddle.pl...
patching ./demo/shapes.pl...
patching ./demo/demo.pl...
patching ./demo/pixel-fx.pl...
patching ./demo/tree.pl...
patching ./demo/lsys.pl...
patching ./demo/steganography.pl...
patching ./check...
patching ./check.sh.in...
patching ./check.sh...
patching ./t/composite.t...
patching ./t/read.t...
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
patching ./t/cgm/read.t...
patching ./t/write.t...
patching ./t/openjp2/read.t...
patching ./t/ping.t...
patching ./t/rad/read.t...
patching ./t/rad/write.t...
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
patching ./t/setattribute.t...
patching ./t/hpgl/read.t...
patching ./t/jng/read.t...
patching ./t/jng/write.t...
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
patching ./t/x11/read.t...
patching ./t/x11/write.t...
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
patching ./t/hdf/read.t...
patching ./t/hdf/write.t...
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
patching ./t/jbig/read.t...
patching ./t/jbig/write.t...
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
patching ./t/getattribute.t...
patching ./t/ps/read.t...
patching ./t/ps/write.t...
patching ./t/blob.t...
patching ./t/wmf/read.t...
patching ./t/bzlib/read.t...
patching ./t/bzlib/write.t...
patching ./t/zlib/read.t...
patching ./t/zlib/write.t...
patching ./t/xfig/read.t...
patching ./t/filter.t...
patching ./t/ttf/read.t...
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
patching ./t/mpeg/read.t...
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
patching ./t/tiff/read.t...
patching ./t/tiff/write.t...
patching ./t/jpeg/read.t...
patching ./t/jpeg/write.t...
patching ./t/png/read.t...
patching ./t/png/write.t...
patching ./t/png/read-16.t...
patching ./t/png/write-16.t...
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
patching ./t/fpx/read.t...
patching ./t/fpx/write.t...
/nix/store/qf3mzpvsmkrw963xchbivcci06078n13-builder.sh: line 17: warning: command substitution: ignored null byte in input
patching ./t/montage.t...
Checking if your kit is complete...
Looks good
Warning (mostly harmless): No library found for -lMagickCore-7.Q16HDRI
Warning (mostly harmless): No library found for -lm
Generating a Unix-style Makefile
Writing Makefile for Image::Magick
Writing MYMETA.yml and MYMETA.json
no configure script, doing nothing
building
build flags: SHELL=/nix/store/vfai0jim0db67nk9rd7ziq29jxb5n79n-bash-5.1-p8/bin/bash
cp Magick.pm blib/lib/Image/Magick.pm
AutoSplitting blib/lib/Image/Magick.pm (blib/lib/auto/Image/Magick)
Running Mkbootstrap for Magick ()
chmod 644 "Magick.bs"
"/nix/store/ag052gyghwrip8zlldk5d0qmhx182zq2-perl-5.34.0/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- Magick.bs blib/arch/auto/Image/Magick/Magick.bs 644
"/nix/store/ag052gyghwrip8zlldk5d0qmhx182zq2-perl-5.34.0/bin/perl" "/nix/store/ag052gyghwrip8zlldk5d0qmhx182zq2-perl-5.34.0/lib/perl5/5.34.0/ExtUtils/xsubpp"  -typemap '/nix/store/ag052gyghwrip8zlldk5d0qmhx182zq2-perl-5.34.0/lib/perl5/5.34.0/ExtUtils/typemap' -typemap '/build/Image-Magick-7.0.11/typemap'  Magick.xs > Magick.xsc
mv Magick.xsc Magick.c
cc -c  -I/nix/store/5pfqzf4fsyxqjm3nssbway66lzpzvhs8-imagemagick-7.1.0-20-dev/include/ImageMagick -D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/no-such-path/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/freetype2 -g -O2 -Wall -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16 -O2   -DVERSION=\"7.0.11\" -DXS_VERSION=\"7.0.11\" -fPIC "-I/nix/store/ag052gyghwrip8zlldk5d0qmhx182zq2-perl-5.34.0/lib/perl5/5.34.0/x86_64-linux-thread-multi/CORE"  -D_LARGE_FILES=1 -DHAVE_CONFIG_H Magick.c
rm -f blib/arch/auto/Image/Magick/Magick.so
cc  -L/usr/local/lib -lMagickCore-7.Q16HDRI -shared -O2 -L/nix/store/41dj1v3qz9a5kjncpkxhmq50yg9r24dn-glibc-2.33-62/lib -fstack-protector-strong   Magick.o  -o blib/arch/auto/Image/Magick/Magick.so  \
      \

chmod 755 blib/arch/auto/Image/Magick/Magick.so
Manifying 1 pod document
running tests
check flags: SHELL=/nix/store/vfai0jim0db67nk9rd7ziq29jxb5n79n-bash-5.1-p8/bin/bash VERBOSE=y test
"/nix/store/ag052gyghwrip8zlldk5d0qmhx182zq2-perl-5.34.0/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- Magick.bs blib/arch/auto/Image/Magick/Magick.bs 644
PERL_DL_NONLAZY=1 "/nix/store/ag052gyghwrip8zlldk5d0qmhx182zq2-perl-5.34.0/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/blob.t .......... ok
t/composite.t ..... ok
t/filter.t ........ ok
t/getattribute.t .. ok
t/montage.t ....... Fontconfig error: Cannot load default config file: No such file: (null)
Fontconfig error: No writable cache directories
Fontconfig error: No writable cache directories
Fontconfig error: No writable cache directories
Fontconfig error: No writable cache directories
t/montage.t ....... ok
t/ping.t .......... ok
t/read.t .......... 1/47 Readimage (input_gray_lsb_double.mat): Exception 425: unexpected end-of-file `input_gray_lsb_double.mat' @ error/mat.c/ReadMATImage/1302 at t/subroutines.pl line 317.
t/read.t .......... Failed 1/47 subtests
t/setattribute.t .. ok
t/write.t ......... ok

Test Summary Report
-------------------
t/read.t        (Wstat: 0 Tests: 47 Failed: 1)
  Failed test:  13
Files=9, Tests=273,  2 wallclock secs ( 0.07 usr  0.02 sys +  3.41 cusr  0.57 csys =  4.07 CPU)
Result: FAIL
Failed 1/9 test programs. 1/273 subtests failed.
make: *** [Makefile:1038: test_dynamic] Error 255
builder for '/nix/store/f8ygrk39l0lgy3s8vs5pwlxgs1af09qa-perl5.34.0-Image-Magick-7.0.11-1.drv' failed with exit code 2
error: build of '/nix/store/f8ygrk39l0lgy3s8vs5pwlxgs1af09qa-perl5.34.0-Image-Magick-7.0.11-1.drv' failed
```

### Expected behavior

`perlPackages.ImageMagick` builds successfully.

### Notify maintainers

@bjornfor (Shutter)

### Metadata

```console
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.12, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"nixos, nixos-unstable"`
 - channels(wenzel): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by bjornfor** [comment]
Maybe we can disable the broken test? Or give fontconfig a writable cache dir somehow?

**Comment by unclechu** [comment]
It seems fontconfig is not the only problem here:
> t/read.t .......... 1/47 Readimage (input_gray_lsb_double.mat): Exception 425: unexpected end-of-file `input_gray_lsb_double.mat' @ error/mat.c/ReadMATImage/1302 at t/subroutines.pl line 317.

**Comment by wamserma** [comment]
Bisected to this commit: https://github.com/NixOS/nixpkgs/commit/918dbfa0c646cd1d8d504919110ce374e0d109ef
ping @oxzi @erictapen

**Comment by wentasah** [comment]
The same thing happens also on nixos-unstable.

**Comment by bjornfor** [comment]
Should be fixed on master and release-21.11. Closing. (It might take a while for the channels to update.)

**Comment by unclechu** [comment]
While it’s not yet in the nixos-21.11 channel I have this as a temporary fix (an overlay):
``` nix
self: super:
let
  imagemagick = super.imagemagick.overrideAttrs (_: {
    # Picked from https://github.com/NixOS/nixpkgs/pull/156920
    # It’s already merged to “release-21.11”, so it should be soon available in one of the next
    # releases of “nixos-21.11” channel. TODO: Remove this overlay when it’s not needed anymore.
    patches = [
      # fix a type confusion bug introduced in 7.1.0-20 with commit 075565e93c71bcaaabf0ce70b7d1060bccdf0020
      (super.fetchpatch {
        url = "https://github.com/ImageMagick/ImageMagick/commit/62845d5672eca4446b952dd0ab2e3e0dab0309d4.patch";
        sha256 = "1kni5i8b5hl69niypidm90mhir8cafi6r9i857fxdlv045h3dg4p";
      })
    ];
  });

  perlPackages = super.perlPackages.override {
    pkgs = super // { inherit imagemagick; };
  };
in
{
  shutter = super.shutter.override {
    inherit imagemagick perlPackages;
  };
}
```
